File: java/revenj-storage/src/main/java/org/revenj/storage/S3Repository.java
package org.revenj.storage;

import java.io.InputStream;
import java.util.Map;
import java.util.concurrent.Future;

public interface S3Repository {
	Future<InputStream> get(String bucket, String key);

	default Future<InputStream> get(S3 s3) {
		return get(s3.getBucket(), s3.getKey());
	}

	Future<Void> upload(
			String bucket,
			String key,
			InputStream stream,
			long length,
			Map<String, String> metadata);

	Future<Void> delete(String bucket, String key);

	default Future<Void> delete(S3 s3) {
		return delete(s3.getBucket(), s3.getKey());
	}
}

File: scala/revenj-storage/src/main/scala/net/revenj/storage/S3Repository.scala
package net.revenj.storage

import software.amazon.awssdk.services.s3.model.DeleteObjectResponse

import java.io.InputStream
import scala.concurrent.Future

trait S3Repository {
	def get(bucket: String, key: String): Future[InputStream]

	def get(s3: S3): Future[InputStream] = get(s3.bucket, s3.key)

	def upload(stream: InputStream, length: Long): Future[S3]

	def upload(
		bucket: String,
		key: String,
		stream: InputStream,
		length: Long,
		name: Option[String] = None,
		mimeType: Option[String] = None,
		metadata: Map[String, String] = Map.empty,
		tags: Map[String, String] = Map.empty): Future[S3]

	def delete(bucket: String, key: String): Future[DeleteObjectResponse]

	def delete(s3: S3): Future[DeleteObjectResponse] = delete(s3.bucket, s3.key)
}